File: dotnet-solution/Utils/ClipCollision.cs
using System;
using System.Collections.Generic;
using System.Linq;
using SpliceIt.Models;

namespace SpliceIt.Utils;

/// <summary>
/// Non-overlap rules for clips sharing a track.
/// Direct port of clipCollision.ts so drag, trim and paste behave identically
/// in both front-ends.
/// </summary>
public static class ClipCollision
{
    private const double Epsilon = 0.001;   // 1 ms — lets clips abut exactly
    private const double BoundsEpsilon = 0.002;

    public static bool Overlaps(
        double start, double duration, IEnumerable<AudioClip> clips, string? ignoreClipId = null)
    {
        double end = start + duration;
        return clips.Any(c =>
        {
            if (ignoreClipId is not null && c.Id == ignoreClipId) return false;
            double cStart = c.TimelineStartSeconds;
            double cEnd = c.TimelineStartSeconds + c.ClipDurationSeconds;
            return start < cEnd - Epsilon && end > cStart + Epsilon;
        });
    }

    /// <summary>
    /// Hard limits a clip may slide or trim to without colliding with a
    /// neighbour. RightBound is PositiveInfinity when nothing follows.
    /// </summary>
    public static (double LeftBound, double RightBound) GetMovementBounds(
        string clipId, double initialStart, double initialDuration, IEnumerable<AudioClip> trackClips)
    {
        var others = trackClips.Where(c => c.Id != clipId).ToList();

        var leftClips = others
            .Where(c => c.TimelineStartSeconds + c.ClipDurationSeconds <= initialStart + BoundsEpsilon)
            .ToList();
        double leftBound = leftClips.Count > 0
            ? leftClips.Max(c => c.TimelineStartSeconds + c.ClipDurationSeconds)
            : 0.0;

        var rightClips = others
            .Where(c => c.TimelineStartSeconds >= initialStart + initialDuration - BoundsEpsilon)
            .ToList();
        double rightBound = rightClips.Count > 0
            ? rightClips.Min(c => c.TimelineStartSeconds)
            : double.PositiveInfinity;

        return (leftBound, rightBound);
    }

    /// <summary>Earliest free slot at or after preferredStart that fits duration.</summary>
    public static double FindNextAvailableSlot(
        double preferredStart, double duration, IEnumerable<AudioClip> clips, string? ignoreClipId = null)
    {
        var sorted = clips
            .Where(c => ignoreClipId is null || c.Id != ignoreClipId)
            .OrderBy(c => c.TimelineStartSeconds)
            .ToList();

        if (sorted.Count == 0) return Math.Max(0, preferredStart);

        if (preferredStart >= 0 &&
            preferredStart + duration <= sorted[0].TimelineStartSeconds + Epsilon)
        {
            return preferredStart;
        }

        double candidate = Math.Max(0, preferredStart);

        for (int i = 0; i < sorted.Count; i++)
        {
            var cur = sorted[i];
            double curEnd = cur.TimelineStartSeconds + cur.ClipDurationSeconds;

            if (candidate < curEnd - Epsilon &&
                candidate + duration > cur.TimelineStartSeconds + Epsilon)
            {
                candidate = curEnd;
            }

            double nextStart = i + 1 < sorted.Count
                ? sorted[i + 1].TimelineStartSeconds
                : double.PositiveInfinity;

            if (candidate + duration <= nextStart + Epsilon) return candidate;
        }

        return candidate;
    }

    /// <summary>Rounds to the nearest grid division when snapping is on.</summary>
    public static double Snap(double seconds, bool snapEnabled, double gridSize)
    {
        if (!snapEnabled || gridSize <= 0) return Math.Max(0, seconds);
        return Math.Max(0, Math.Round(seconds / gridSize) * gridSize);
    }
}

File: src/utils/clipCollision.ts
import { AudioClipModel } from '../types';

/**
 * Checks if a proposed range [start, start + duration] overlaps with any other clip on the track.
 */
export function checkClipOverlap(
  start: number,
  duration: number,
  clips: AudioClipModel[],
  ignoreClipId?: string
): boolean {
  const end = start + duration;
  const epsilon = 0.001; // 1ms tolerance to allow perfectly abutting clips
  return clips.some((c) => {
    if (ignoreClipId && c.id === ignoreClipId) return false;
    const cStart = c.timelineStart;
    const cEnd = c.timelineStart + c.clipDuration;
    return start < cEnd - epsilon && end > cStart + epsilon;
  });
}

/**
 * Calculates strict boundaries [leftBound, rightBound] for a clip to slide/trim
 * without overlapping any other clip on the track.
 */
export function getClipMovementBounds(
  clipId: string,
  initialStart: number,
  initialDuration: number,
  trackClips: AudioClipModel[]
): { leftBound: number; rightBound: number } {
  const otherClips = trackClips.filter((c) => c.id !== clipId);
  const epsilon = 0.002;

  // Clips strictly to the left of this clip's current position
  const leftClips = otherClips.filter((c) => c.timelineStart + c.clipDuration <= initialStart + epsilon);
  const leftBound = leftClips.length > 0
    ? Math.max(...leftClips.map((c) => c.timelineStart + c.clipDuration))
    : 0;

  // Clips strictly to the right of this clip's current position
  const rightClips = otherClips.filter((c) => c.timelineStart >= initialStart + initialDuration - epsilon);
  const rightBound = rightClips.length > 0
    ? Math.min(...rightClips.map((c) => c.timelineStart))
    : Infinity;

  return { leftBound, rightBound };
}

/**
 * Finds the earliest non-overlapping timeline position (>= preferredStart) for a clip of given duration.
 */
export function findNextAvailableSlot(
  preferredStart: number,
  duration: number,
  clips: AudioClipModel[],
  ignoreClipId?: string
): number {
  const sorted = clips
    .filter((c) => !ignoreClipId || c.id !== ignoreClipId)
    .sort((a, b) => a.timelineStart - b.timelineStart);

  if (sorted.length === 0) {
    return Math.max(0, preferredStart);
  }

  // Check if preferredStart fits before first clip
  if (preferredStart + duration <= sorted[0].timelineStart + 0.001 && preferredStart >= 0) {
    return preferredStart;
  }

  // Check gaps between clips starting from or after preferredStart
  let candidate = Math.max(0, preferredStart);

  for (let i = 0; i < sorted.length; i++) {
    const cur = sorted[i];
    const curEnd = cur.timelineStart + cur.clipDuration;

    // If candidate falls inside this clip, advance candidate to end of this clip
    if (candidate < curEnd - 0.001 && candidate + duration > cur.timelineStart + 0.001) {
      candidate = curEnd;
    }

    // Check gap after this clip and before next clip
    const nextStart = i + 1 < sorted.length ? sorted[i + 1].timelineStart : Infinity;
    if (candidate + duration <= nextStart + 0.001) {
      return candidate;
    }
  }

  return candidate;
}
